src/clj/puppetlabs/puppetserver/certificate_authority.clj

```diff
@@ -29,6 +29,54 @@
             [puppetlabs.puppetserver.shell-utils :as shell-utils]
             [puppetlabs.i18n.core :as i18n]))
 
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;;; Public utilities
+
+;; Pattern is one or more digit followed by time unit
+(def digits-with-unit-pattern #"(\d+)(y|d|h|m|s)")
+(def repeated-digits-with-unit-pattern #"((\d+)(y|d|h|m|s))+")
+
+(defn duration-string?
+  "Returns true if string is formatted with duration string pairs only, otherwise returns nil.
+   Ignores whitespace."
+  [maybe-duration-string]
+  (when (string? maybe-duration-string)
+    (let [no-whitespace-string (clojure.string/replace maybe-duration-string #" " "")]
+      (some? (re-matches repeated-digits-with-unit-pattern no-whitespace-string)))))
+
+(defn duration-str->sec
+  "Converts a string containing any combination of duration string pairs in the format '<num>y' '<num>d' '<num>m' '<num>h' '<num>s'
+   to a total number of seconds.
+   nil is returned if the input is not a string or not a string containing any valid duration string pairs."
+  [string-input]
+  (when (duration-string? string-input)
+    (let [pattern-matcher (re-matcher digits-with-unit-pattern string-input)
+          first-match (re-find pattern-matcher)]
+      (loop [[_match-str digits unit] first-match
+             running-total 0]
+        (let [unit-in-seconds (case unit
+                                "y" 31536000 ;; 365 day year, not a real year
+                                "d" 86400
+                                "h" 3600
+                                "m" 60
+                                "s" 1)
+              total-seconds (+ running-total (* (Integer/parseInt digits) unit-in-seconds))
+              next-match (re-find pattern-matcher)]
+          (if (some? next-match)
+            (recur next-match total-seconds)
+            total-seconds))))))
+
+(defn get-ca-ttl
+  "Returns ca-ttl value as an integer. If a value is set in certificate-authority that value is returned.
+   Otherwise puppet config setting is returned"
+  [puppetserver certificate-authority]
+  (when (not-empty (select-keys certificate-authority [:ca-ttl]))
+    (update certificate-authority :ca-ttl duration-str->sec)
```

Comment:
This statement by itself won't change the value in `certificate-authority`, so you will need to restructure this.  Perhaps something like:
```
(let [ca-config-value (duration-str->sec (:ca-ttl certificate-authority))
        puppet-config-value (:ca-ttl puppetserver)]
        (when (and ca-config-value puppet-config-value)
            (log/warn (i18n/trs) "Detected ca-ttl setting in CA config which will take precedence over puppet.conf setting"))
        (or ca-config-value puppet-config-value))
```